src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/SyntaxNodeExtensions.cs

```diff
@@ -328,5 +329,60 @@ public static int GetTrailingWhitespaceLength(this SyntaxNode node, FormattingCo
     }
 
     public static SyntaxNode? FindInnermostNode(this SyntaxNode node, int index, bool includeWhitespace = false)
-        => node.FindToken(index, includeWhitespace)?.Parent;
+    {
+        var token = node.FindToken(index, includeWhitespace);
+
+        // If the index is EOF but the node has index-1,
+        // then try to get a token to the left of the index.
+        // patterns like
+        // <button></button>$$
+        // should get the button node instead of the razor document (which is the parent
+        // of the EOF token)
+        if (token.Kind == SyntaxKind.EndOfFile && node.Span.Contains(index - 1))
+        {
+            token = node.FindToken(index - 1, includeWhitespace);
+        }
+
+        return token.Parent;
+    }
+
+    public static SyntaxNode? FindNode(this SyntaxNode @this, Language.Syntax.TextSpan span, bool includeWhitespace = false, bool getInnermostNodeForTie = false)
```

Comment:
shamelessly copied from Roslyn.

Reply:
This looks scary, and seems like it only has one caller, which is surprising and doesn't fill me with joy, but the fact that its copied from Roslyn means I'm going to assume its fine :)

Reply:
I should probably add a lot of tests

Reply:
I was wondering about that, and also wondering if this should move down to the compiler, next to `FindToken`. I think @333fred has mentioned thinking about adding something that dealt with ties, but couldn't see any usage that required spans, so left it out.

Reply:
We can certainly move this down to the compiler, but yes the initial `FindToken` method was all we thought we needed in the design meeting a few months ago. If we find that need `FindNode` as well, I'm perfectly fine with adding it.

Reply:
IIRC, `FindToken` was considered table stakes and that we'd build whatever helpers we needed based on that and then push APIs down to the compiler that were particularly useful. In Roslyn, `SyntaxNode` didn't have a `FindNode` API for many years (or `Ancestors`, `Descendants`, FirstAncestorOrSelf`, etc.) and were only added when it was recognized that they were essential. I'm happy to *not* move this to the compiler unless it's super useful for other purposes.